dali/kernels/imgproc/convolution/separable_convolution_gpu.h

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_SEPARABLE_CONVOLUTION_GPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_SEPARABLE_CONVOLUTION_GPU_H_
+
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/convolution_gpu.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Apply convolution in all spatial axes, starting from the innermost to outermost.
+ *        If channel axis is pressent, the convolution is not applied there.
+ *        If it is marqed as sequence, the first data axis is considered as temporal axis
+ *
+ * Sequence convolutions require one less window to be specified
+ *
+ * `W` is currently assumed to be float
+ *
+ * `windows` and `scales` are specified per axis.
+ *
+ * Specialized for 1, 2 or 3 axes, to not go overboard with TMP for generic solutions
+ *
+ * Here be boilerplate.
+ */
+template <typename Out, typename In, typename W, int axes, bool has_channels = false,
+          bool is_sequence = false>
+struct SeparableConvolutionGpu;
+
+template <typename Out, typename In, typename W, bool has_channels, bool is_sequence>
+struct SeparableConvolutionGpu<Out, In, W, 1, has_channels, is_sequence> {
+  static constexpr int axes = 1;
+  static constexpr int sequence_axes = static_cast<int>(is_sequence);
+  static constexpr int channel_axes = static_cast<int>(has_channels);
+  static constexpr int ndim = sequence_axes + axes + channel_axes;
```

Comment:
It seems to be the same for all specialization, can we extract this?

Reply:
Do you have any idea how to make it better? I can maybe put it in some
```
template <int axes, bool is_sequence, bool has_channels>
struct calc_ndim {
	static constexpr int ndim = sequence_axes + axes + channel_axes;
};
```
and skip the sequence_axes & channel_axes members, but I'm not sure how much better is it.

Reply:
Just an observation, if you don't see any better way leave it as it is.

Reply:
maybe a constexpr function can also work.

Reply:
Hmm, actually 3 of those member variables are used directly in some kernel parameters definitions or calls. So I will just leave it.